### ISSUE TITLE
[Feature Request] Official Cavalcade Support

### ISSUE THREAD

**Issue by larssn** [Prospress/action-scheduler#159]
Hello
I'd like to request an official way to use this plugin with Cavalcade.

Since WooCommerce adopted this plugin; the tasks that were normally run by Cavalcade have now stopped running, and have to be run manually.

Any reason why the plugin explicitly requires WP_CRON in order to function?

**Comment by thenbrent** [comment]
> the tasks that were normally run by Cavalcade have now stopped running, and have to be run manually.

It was my understanding that Cavalcade was a drop-in replacement for WP Cron, so Action Scheduler should _just work_:tm: with it. It sounds like that's not happening though, so there must be a compatibility issue between the two.

For reference, Action Scheduler sets itself up to run on WP Cron in `ActionScheduler_QueueRunner::init()` here: https://github.com/Prospress/action-scheduler/blob/2.0.0-beta-2/classes/ActionScheduler_QueueRunner.php#L57

My first guess around why things aren't working is that there is some issue with load order of the two codebases resulting in `wp_schedule_event()` not having been overloaded by Cavalcade when that code is run. But that's just a guess.

In any case, I don't think a custom runner should be required for Cavalcade, instead a fix for how Action Scheduler's schedules itself with WP Cron and/or for how Cavalcade overrides WP Cron.

> Any reason why the plugin explicitly requires WP_CRON in order to function?

Action Scheduler doesn't _require_ WP Cron, it just uses that by default to initiate queue runners because it's available out-of-the-box on all WordPress sites (unlike Cavalcade, WP CLI or other potential runners).

For more details on that distinction, see https://github.com/Prospress/action-scheduler#i-dont-want-to-use-wp-cron-does-action-scheduler-depend-on-wp-cron

On large sites, I recommend disabling the default WP Cron Queue Runner entirely with [this plugin](https://github.com/prospress/action-scheduler-disable-default-runner) and using exclusively [WP CLI runner](https://github.com/Prospress/action-scheduler#wp-cli) instead.

**Comment by larssn** [comment]
Thanks for the detailed response!

It's good to hear that its _supposed_ to work. The code you linked seems like a perfectly reasonable way to "bootstrap" action scheduler against the existing cron system, which is what makes it very strange.

We **know** that Cavalcade works on our existing sites, because we use it to schedule outgoing SMS'es, and they get fired off without problems. The offending plugin atm is WooCommerce Subscriptions.

I haven't looked into the load order that you mentioned, but it shouldn't be a problem, as Cavalcade overrides said methods as a mu-plugin - so it should bootstrap its stuff first, always.

**Comment by thenbrent** [comment]
> The offending plugin atm is WooCommerce Subscriptions.

Good to know. Can you submit a support ticket about this issue via http://prospress.com/support/

I suspect we'll need to dig deep before we figure out the issue, so support email might be a better approach.

If you can include temp admin creds on the ticket submission, that will be a great help as it will allow me to take a look at the Action Scheduler queues and other system status related things. 

Let me try to pull in @rmccue to this thread to see if his expertise can help provide any clues on where to look before I dive in too.

👋 @rmccue

**Comment by larssn** [comment]
I've made a ticket, however, I'll need to contact the customer in order to provide admin permissions

**Comment by JPry** [comment]
Hi @larssn, 

I wanted to add on to what Brent mentioned earlier. In regards to this:

> Since WooCommerce adopted this plugin; the tasks that were normally run by Cavalcade have now stopped running, and have to be run manually.

Could you clarify what you mean here? WooCommerce core has not officially adopted this library (though it is indeed [being considered](https://github.com/woocommerce/woocommerce/pull/20030)), and Subscriptions has been using this library for a long time now. Additionally, while there is always on-going development for both Action Scheduler and Subscriptions, we haven't (at the time that I'm writing this) released any major updates to either. Subscriptions 2.3 (currently in beta) contains recent code updates to Action Scheduler, but the last time Action Scheduler was updated with Subscriptions was in version `2.2.6`.

In other words, we haven't made any recent changes that would seem likely to cause the sudden changes you're describing. 

I haven't yet set up Cavalcade locally to test with Action Scheduler. With that said, my initial thought about would could possibly be the problem is the Cavalcade runner. Since Cavalcade disables normal WP CRON, it expects the runner to be the replacement when it comes to triggering jobs to run. Can you confirm that the server still has the runner running properly? Are you able to have jobs run correctly when using Cavalcade via WP CLI?

I'm going to try to get this up and running this week to test Cavalcade directly, but if you're able to look into the Cavalcade setup on your side, that would be very helpful as well.

**Comment by larssn** [comment]
> Could you clarify what you mean here? WooCommerce core has not officially adopted this library (though it is indeed being considered), and Subscriptions has been using this library for a long time now.
Sorry, maybe I've been misinformed. I'm sure you're right. I assumed there was some recent development, causing the problem.

Let me give a bit more information. We originally bought the plugin back in 2012, and then never really used it until sometime last year, when one of our customers requested a subscription service. It ran just fine to begin with, but a few months back (probably April), the customer notified us that subscriptions were no longer going out. We update Wordpress on a regular basis, as is expected of us (related?). So for now we run the subscription jobs manually, which takes a while.

> Can you confirm that the server still has the runner running properly? Are you able to have jobs run correctly when using Cavalcade via WP CLI?
We dont use the CLI, but the runner seems to work, as we use Cavalcade for all other jobs; WP_CRON is disabled in `wp-config`. Jetpack has a bunch of jobs that run just fine, as well as our own jobs which send out the occasional SMS. We updated Cavalcade recently, after the problems began, but doing that didn't fix it (nor make it worse).

I'm still waiting for a response from the customer regarding admin access.

**Comment by larssn** [comment]
I've sent the login info to your support department.

**Comment by JPry** [comment]
@larssn Thanks for sending along those details. I took a look around the site, and didn't notice anything that seemed unusual, although it's worth noting that the site seemed to be unusually slow overall.

I set up Cavalcade with my local install today to test how it behaves with Action Scheduler. I followed all of the instructions from the Cavalcade repository. Action Scheduler actions were created as expected, and the queue runner job (normally run by WP Cron) was added to Cavalcade's queue:

```
wp cavalcade jobs                         
+----+------+------------------------------------+---------------------+---------------------+---------+
| id | site | hook                               | start               | nextrun             | status  |
+----+------+------------------------------------+---------------------+---------------------+---------+
| 27 | 1    | action_scheduler_run_queue         | 2018-06-26 15:45:19 | 2018-06-26 18:55:19 | waiting |
| 28 | 1    | wp_privacy_delete_old_export_files | 2018-06-26 15:45:19 | 2018-06-26 19:45:19 | waiting |
| 29 | 1    | wcs_cleanup_big_logs               | 2018-06-26 15:45:19 | 2018-07-03 15:45:19 | waiting |
| 30 | 1    | wp_version_check                   | 2018-06-26 16:47:06 | 2018-06-27 04:47:06 | waiting |
| 31 | 1    | wp_update_plugins                  | 2018-06-26 16:47:06 | 2018-06-27 04:47:06 | waiting |
| 32 | 1    | wp_update_themes                   | 2018-06-26 16:47:06 | 2018-06-27 04:47:06 | waiting |
| 33 | 1    | wp_scheduled_delete                | 2018-06-26 16:47:06 | 2018-06-27 16:47:06 | waiting |
| 34 | 1    | delete_expired_transients          | 2018-06-26 16:47:06 | 2018-06-27 16:47:06 | waiting |
+----+------+------------------------------------+---------------------+---------------------+---------+
```

When I manually ran the job using `wp cavalcade run 27`, I confirmed that scheduled actions executed as expected.

I then installed the Cavalcade Runner and ran a worker. I confirmed that while the worker was running, it successfully triggered the `action_scheduler_run_queue` action every 60 seconds (after catching up to the current time), just as expected. On top of that, any actions that were due to be run were executed properly.

I can only guess at why your jobs aren't running, but I can think of two possibilities:

* It's possible that the `action_scheduler_run_queue` job has been marked as `failed` in the Cavalcade job table. This would prevent the job from executing properly, based on the information available in humanmade/Cavalcade#23. I would recommend looking at the jobs table and confirming the status of the action scheduler job.
* It's possible that the Cavalcade runner isn't actually running, or else it is only running intermittently. I would recommend confirming that the Runner is actually running and that it is able to process jobs.

Since my testing indicates that Cavalcade should be working with Action Scheduler without any issues, I'm going to go ahead and close this issue. You are welcome to post any additional questions or feedback here. While I'm willing to answer questions, I'm not sure if I'll be able to help troubleshoot why Cavalcade isn't working on your particular site. I don't have much experience with Cavalcade other than the tests I've run on my own local site.
